File: mcp/lib/src/workspace.dart
import 'dart:io';

import 'package:workout_log_core/workout_log_core.dart';

/// The session folder as a real directory (ADR-007's desktop case).
///
/// Only the byte-level plumbing lives here: every rule about what the bytes
/// mean stays in `workout_log_core` (ADR-004).
class DirectoryStorage implements SessionStorage {
  DirectoryStorage(this.directory);

  final Directory directory;

  @override
  String get label => directory.path;

  @override
  Future<List<String>> listIds() async {
    if (!directory.existsSync()) return const [];
    return directory
        .listSync()
        .whereType<File>()
        .map((f) => f.uri.pathSegments.last)
        .toList();
  }

  @override
  Future<String> read(String id) => _file(id).readAsString();

  @override
  Future<void> write(String id, String contents) async {
    await directory.create(recursive: true);
    // Write-then-rename, as the app does: a crash mid-write leaves the previous
    // session intact rather than a truncated file.
    final temporary = File('${_file(id).path}.tmp');
    await temporary.writeAsString(contents, flush: true);
    await temporary.rename(_file(id).path);
  }

  @override
  Future<void> delete(String id) async {
    final file = _file(id);
    if (file.existsSync()) await file.delete();
  }

  File _file(String id) {
    if (id.contains('/') || id.contains('..')) {
      throw ArgumentError.value(id, 'id', 'must be a bare file name');
    }
    return File('${directory.path}/$id');
  }
}

class WorkspaceException implements Exception {
  WorkspaceException(this.message);

  final String message;

  @override
  String toString() => message;
}

/// The repository the server works against: the session folder plus the two
/// hand-maintained reference files beside it.
class Workspace {
  Workspace({required this.root, required Directory data})
      : sessions = SessionStore(DirectoryStorage(data)),
        dataDirectory = data;

  /// Resolution order matches the app's (ADR-007): an explicit path, then
  /// `$WORKOUTLOG_DATA`, then `data/` under whichever ancestor holds cycles.json.
  factory Workspace.open({
    String? root,
    String? data,
    Map<String, String> environment = const {},
  }) {
    final repo = Directory(root ?? _findRoot()?.path ?? Directory.current.path);
    final dataPath = data ?? environment['WORKOUTLOG_DATA'];
    return Workspace(
      root: repo,
      data: Directory(dataPath ?? '${repo.path}/data'),
    );
  }

  static Directory? _findRoot() {
    var dir = Directory.current.absolute;
    while (true) {
      if (File('${dir.path}/cycles.json').existsSync()) return dir;
      final parent = dir.parent;
      if (parent.path == dir.path) return null;
      dir = parent;
    }
  }

  final Directory root;
  final Directory dataDirectory;
  final SessionStore sessions;

  String get label => dataDirectory.path;

  File get catalogueFile => File('${root.path}/exercises.json');

  File get cyclesFile => File('${root.path}/cycles.json');

  Catalogue loadCatalogue() =>
      SessionCoding.decodeCatalogue(_readReference(catalogueFile));

  CycleCatalogue loadCycles() =>
      CycleCatalogue.fromJson(jsonMap(_readReference(cyclesFile)));

  void saveCatalogue(Catalogue catalogue) =>
      _writeReference(catalogueFile, SessionCoding.encodeJson(catalogue.toJson()));

  void saveCycles(CycleCatalogue cycles) =>
      _writeReference(cyclesFile, SessionCoding.encodeJson(cycles.toJson()));

  /// Every session in the archive, newest last, with unreadable files reported
  /// rather than thrown: a corrupted file costs one session, never the archive.
  Future<LoadResult> loadAll({String? from, String? to}) async {
    final result = await sessions.loadAll();
    final kept = [
      for (final session in result.sessions)
        if (_inRange(session.date, from, to)) session,
    ]..sort((a, b) => a.date.compareTo(b.date));
    return LoadResult(kept, result.failures);
  }

  static bool _inRange(String date, String? from, String? to) =>
      (from == null || date.compareTo(from) >= 0) &&
      (to == null || date.compareTo(to) <= 0);

  String _readReference(File file) {
    if (!file.existsSync()) {
      throw WorkspaceException(
        'no ${file.uri.pathSegments.last} under ${root.path} — point the server '
        'at the repository with --repo',
      );
    }
    return file.readAsStringSync();
  }

  /// Writing a reference file also refreshes the copy bundled into the Flutter
  /// app, which is what `tool/sync_assets.dart` does by hand; without it the
  /// asset-sync test starts failing the moment this server edits a catalogue.
  void _writeReference(File file, String contents) {
    file.writeAsStringSync(contents);
    final bundled = File(
      '${root.path}/flutter/assets/data/${file.uri.pathSegments.last}',
    );
    if (bundled.parent.existsSync()) bundled.writeAsStringSync(contents);
  }
}

File: crates/workout-log-mcp/src/workspace.rs
//! The repository the server is pointed at: the session folder plus the two
//! hand-maintained reference files beside it.

use std::fs;
use std::path::{Path, PathBuf};

use workout_log_core::coding::encode_json;
use workout_log_core::io::session_storage::{SessionStorage, StorageError, StorageResult};
use workout_log_core::io::session_store::SessionStore;
use workout_log_core::{Catalogue, CycleCatalogue, Session};

#[derive(Debug, Clone, PartialEq, Eq, thiserror::Error)]
#[error("{0}")]
pub struct WorkspaceError(pub String);

pub struct DirectoryStorage {
    directory: PathBuf,
}

impl DirectoryStorage {
    pub fn new(directory: PathBuf) -> Self {
        Self { directory }
    }

    fn file(&self, id: &str) -> StorageResult<PathBuf> {
        if id.contains('/') || id.contains("..") {
            return Err(StorageError(format!(
                "Invalid argument(s) (id): must be a bare file name: \"{id}\""
            )));
        }
        Ok(self.directory.join(id))
    }
}

impl SessionStorage for DirectoryStorage {
    fn list_ids(&self) -> StorageResult<Vec<String>> {
        if !self.directory.is_dir() {
            return Ok(Vec::new());
        }
        let entries = fs::read_dir(&self.directory).map_err(|e| StorageError(e.to_string()))?;
        Ok(entries
            .filter_map(Result::ok)
            .filter(|e| e.path().is_file())
            .map(|e| e.file_name().to_string_lossy().into_owned())
            .collect())
    }

    fn read(&self, id: &str) -> StorageResult<String> {
        fs::read_to_string(self.file(id)?).map_err(|e| StorageError(e.to_string()))
    }

    /// Write-then-rename: a crash mid-write leaves the previous session intact
    /// rather than a truncated file.
    fn write(&mut self, id: &str, contents: &str) -> StorageResult<()> {
        let path = self.file(id)?;
        fs::create_dir_all(&self.directory).map_err(|e| StorageError(e.to_string()))?;
        let temporary = path.with_extension("json.tmp");
        fs::write(&temporary, contents).map_err(|e| StorageError(e.to_string()))?;
        fs::rename(&temporary, &path).map_err(|e| StorageError(e.to_string()))
    }

    fn delete(&mut self, id: &str) -> StorageResult<()> {
        let path = self.file(id)?;
        if path.exists() {
            fs::remove_file(path).map_err(|e| StorageError(e.to_string()))?;
        }
        Ok(())
    }

    fn label(&self) -> String {
        self.directory.display().to_string()
    }
}

pub struct Workspace {
    pub root: PathBuf,
    pub data_directory: PathBuf,
    pub sessions: SessionStore,
}

pub struct LoadOutcome {
    pub sessions: Vec<Session>,
    pub failures: Vec<(String, String)>,
}

impl Workspace {
    /// `--repo` wins, else the nearest ancestor holding `cycles.json`, else the
    /// working directory. `--data` wins over `$WORKOUTLOG_DATA`, else `<root>/data`.
    pub fn open(
        root: Option<PathBuf>,
        data: Option<PathBuf>,
        from_environment: Option<String>,
    ) -> Self {
        let root = root
            .or_else(find_root)
            .unwrap_or_else(|| PathBuf::from("."));
        let data_directory = data
            .or_else(|| {
                from_environment
                    .filter(|v| !v.is_empty())
                    .map(PathBuf::from)
            })
            .unwrap_or_else(|| root.join("data"));
        Self {
            sessions: SessionStore::new(Box::new(DirectoryStorage::new(data_directory.clone()))),
            root,
            data_directory,
        }
    }

    pub fn label(&self) -> String {
        self.data_directory.display().to_string()
    }

    pub fn catalogue_file(&self) -> PathBuf {
        self.root.join("exercises.json")
    }

    pub fn cycles_file(&self) -> PathBuf {
        self.root.join("cycles.json")
    }

    pub fn load_catalogue(&self) -> Result<Catalogue, WorkspaceError> {
        let text = self.read_reference(&self.catalogue_file())?;
        serde_json::from_str(&text).map_err(|e| WorkspaceError(e.to_string()))
    }

    pub fn load_cycles(&self) -> Result<CycleCatalogue, WorkspaceError> {
        let text = self.read_reference(&self.cycles_file())?;
        serde_json::from_str(&text).map_err(|e| WorkspaceError(e.to_string()))
    }

    pub fn save_catalogue(&self, catalogue: &Catalogue) -> Result<(), WorkspaceError> {
        self.write_reference(&self.catalogue_file(), catalogue)
    }

    pub fn save_cycles(&self, cycles: &CycleCatalogue) -> Result<(), WorkspaceError> {
        self.write_reference(&self.cycles_file(), cycles)
    }

    /// Range bounds are inclusive and compared as text, which is what ISO dates
    /// are for. Parse failures are reported, never thrown away.
    pub fn load_all(&self, from: Option<&str>, to: Option<&str>) -> LoadOutcome {
        let loaded = match self.sessions.load_all() {
            Ok(loaded) => loaded,
            Err(error) => {
                return LoadOutcome {
                    sessions: Vec::new(),
                    failures: vec![(self.label(), error.to_string())],
                };
            }
        };
        let mut sessions: Vec<Session> = loaded
            .sessions
            .into_iter()
            .filter(|s| in_range(&s.date, from, to))
            .collect();
        sessions.sort_by(|a, b| a.date.cmp(&b.date));
        LoadOutcome {
            sessions,
            failures: loaded
                .failures
                .into_iter()
                .map(|f| (f.id, f.error))
                .collect(),
        }
    }

    fn read_reference(&self, path: &Path) -> Result<String, WorkspaceError> {
        fs::read_to_string(path).map_err(|_| {
            WorkspaceError(format!(
                "no {} under {} — point the server at the repository with --repo",
                path.file_name().unwrap_or_default().to_string_lossy(),
                self.root.display()
            ))
        })
    }

    fn write_reference<T: serde::Serialize>(
        &self,
        path: &Path,
        value: &T,
    ) -> Result<(), WorkspaceError> {
        let text =
            encode_json(&serde_json::to_value(value).map_err(|e| WorkspaceError(e.to_string()))?);
        let temporary = path.with_extension("json.tmp");
        fs::write(&temporary, text).map_err(|e| WorkspaceError(e.to_string()))?;
        fs::rename(&temporary, path).map_err(|e| WorkspaceError(e.to_string()))
    }
}

fn in_range(date: &str, from: Option<&str>, to: Option<&str>) -> bool {
    from.is_none_or(|f| date >= f) && to.is_none_or(|t| date <= t)
}

/// The nearest ancestor of the working directory that holds `cycles.json`.
pub fn find_root() -> Option<PathBuf> {
    std::env::current_dir()
        .ok()?
        .ancestors()
        .find(|dir| dir.join("cycles.json").is_file())
        .map(Path::to_path_buf)
}
